// ngx-papaparse/bundles/ngx-papaparse.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('papaparse/papaparse.min.js')) :
    typeof define === 'function' && define.amd ? define('ngx-papaparse', ['exports', '@angular/core', 'papaparse/papaparse.min.js'], factory) :
    (factory((global['ngx-papaparse'] = {}),global.ng.core,null));
}(this, (function (exports,core,lib) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var Papa = /** @class */ (function () {
        function Papa(config) {
            this.config = config;
            this._papa = lib;
            if (!this.config) {
                this.config = {};
            }
        }
        /**
         * Parse CSV to an array
         * @param {?} csv
         * @param {?=} config
         * @return {?}
         */
        Papa.prototype.parse = /**
         * Parse CSV to an array
         * @param {?} csv
         * @param {?=} config
         * @return {?}
         */
            function (csv, config) {
                if (config) {
                    if (config.worker === true) {
                        if (this.config.scriptPath) {
                            this._papa.SCRIPT_PATH = this.config.scriptPath;
                        }
                        else {
                            throw new Error('When using workers, the workerScriptPath must be defined in global' +
                                ' papaparse configuration. See' +
                                ' https://alberthaff.dk/projects/ngx-papaparse/docs/v3/parsing-csv/using-serviceworkers' +
                                ' for more information.');
                        }
                    }
                }
                return this._papa.parse(csv, config);
            };
        /**
         * Convert an array into CSV
         * @param {?} data
         * @param {?=} config
         * @return {?}
         */
        Papa.prototype.unparse = /**
         * Convert an array into CSV
         * @param {?} data
         * @param {?=} config
         * @return {?}
         */
            function (data, config) {
                return this._papa.unparse(data, config);
            };
        /**
         * Set the size in bytes of each file chunk.
         * Used when streaming files obtained from the DOM that
         * exist on the local computer. Default 10 MB.
         * @param {?} value
         * @return {?}
         */
        Papa.prototype.setLocalChunkSize = /**
         * Set the size in bytes of each file chunk.
         * Used when streaming files obtained from the DOM that
         * exist on the local computer. Default 10 MB.
         * @param {?} value
         * @return {?}
         */
            function (value) {
                this._papa.LocalChunkSize = value;
            };
        /**
         * Set the size in bytes of each remote file chunk.
         * Used when streaming remote files. Default 5 MB.
         * @param {?} value
         * @return {?}
         */
        Papa.prototype.setRemoteChunkSize = /**
         * Set the size in bytes of each remote file chunk.
         * Used when streaming remote files. Default 5 MB.
         * @param {?} value
         * @return {?}
         */
            function (value) {
                this._papa.RemoteChunkSize = value;
            };
        /**
         * Set the delimiter used when it is left unspecified and cannot be detected automatically. Default is comma.
         * @param {?} value
         * @return {?}
         */
        Papa.prototype.setDefaultDelimiter = /**
         * Set the delimiter used when it is left unspecified and cannot be detected automatically. Default is comma.
         * @param {?} value
         * @return {?}
         */
            function (value) {
                this._papa.DefaultDelimiter = value;
            };
        Object.defineProperty(Papa.prototype, "badDelimiters", {
            /**
             * An array of characters that are not allowed as delimiters.
             */
            get: /**
             * An array of characters that are not allowed as delimiters.
             * @return {?}
             */ function () {
                return this._papa.BAD_DELIMITERS;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Papa.prototype, "recordSeperator", {
            /**
             * The true delimiter. Invisible. ASCII code 30.
             * Should be doing the job we strangely rely upon commas and tabs for.
             */
            get: /**
             * The true delimiter. Invisible. ASCII code 30.
             * Should be doing the job we strangely rely upon commas and tabs for.
             * @return {?}
             */ function () {
                return this._papa.RECORD_SEP;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Papa.prototype, "unitSeperator", {
            /**
             * Also sometimes used as a delimiting character. ASCII code 31.
             */
            get: /**
             * Also sometimes used as a delimiting character. ASCII code 31.
             * @return {?}
             */ function () {
                return this._papa.UNIT_SEP;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Papa.prototype, "workersSupported", {
            /**
             * Whether or not the browser supports HTML5 Web Workers.
             * If false, worker: true will have no effect.
             */
            get: /**
             * Whether or not the browser supports HTML5 Web Workers.
             * If false, worker: true will have no effect.
             * @return {?}
             */ function () {
                return this._papa.WORKERS_SUPPORTED;
            },
            enumerable: true,
            configurable: true
        });
        Papa.decorators = [
            { type: core.Injectable },
        ];
        /** @nocollapse */
        Papa.ctorParameters = function () {
            return [
                { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: ['PapaParseGlobalConfig',] }] }
            ];
        };
        return Papa;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var PapaParseModule = /** @class */ (function () {
        function PapaParseModule() {
        }
        PapaParseModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [],
                        providers: [
                            Papa
                        ]
                    },] },
        ];
        return PapaParseModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.Papa = Papa;
    exports.PapaParseModule = PapaParseModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
